File: Trees/230. 二叉搜索树中第K小的元素/230. 二叉搜索树中第K小的元素.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# preorder中序解法，差解
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        res = []
        q = collections.deque()
        q.append(root)

        while q:
            sz = len(q)
            for i in range(sz):
                node = q.pop()

                if node.left : q.append(node.left)
                if node.right : q.append(node.right)

                res.append(node.val)
        
        res.sort()

        return res[k-1]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# inorder中序解法，較佳
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        res = []
        stack = []
        while root or stack:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()

            res.append(root.val)
            # k -= 1
            # if k == 0:
            #     return root.val

            root = root.right
        
        return res[k - 1]

File: Trees/230. 二叉搜索树中第K小的元素/230. 二叉搜索树中第K小的元素.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// preorder中序解法，差解
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        // init
        vector<int> all_v;
        queue<TreeNode*> q;
        q.push(root);

        while(!q.empty()){
            int sz = q.size();
            for(int i = 0; i < sz; i++){
                // node為q最前的一個元素
                TreeNode* node = q.front();
                q.pop();

                if(node->left != nullptr) q.push(node->left);
                if(node->right != nullptr) q.push(node->right);

                all_v.push_back(node->val);
            }
        }

        sort(all_v.begin(), all_v.begin() + all_v.size());

        return all_v[k - 1];
    }
};

// inorder中序解法，較佳
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> res;
        if (root == nullptr) return 0;
        stack<TreeNode *> s;

        while (root != nullptr || !s.empty()) {
            while (root != nullptr) {
                s.push(root);
                root = root->left;
            } 

            TreeNode *node = s.top();
            s.pop();
            res.push_back(node->val);
            root = node->right;
        }

        return res[k - 1];
    }
};